scalalib/src/mill/scalalib/publish/SonatypeCentralPublisher.scala

```diff
@@ -0,0 +1,84 @@
+package mill.scalalib.publish
+
+import com.lumidion.sonatype.central.client.core.{
+  DeploymentName,
+  PublishingType,
+  SonatypeCredentials
+}
+import com.lumidion.sonatype.central.client.requests.SyncSonatypeClient
+import mill.api.Logger
+
+import java.io.FileOutputStream
+import java.util.jar.JarOutputStream
+import java.util.zip.ZipEntry
+
+class SonatypeCentralPublisher(
+    credentials: SonatypeCredentials,
+    signed: Boolean,
+    gpgArgs: Seq[String],
+    readTimeout: Int,
+    connectTimeout: Int,
+    log: Logger,
+    workspace: os.Path,
+    env: Map[String, String],
+    awaitTimeout: Int
+) extends SonatypeHelpers {
+  private val sonatypeCentralClient =
+    new SyncSonatypeClient(credentials, readTimeout = readTimeout, connectTimeout = connectTimeout)
+
+  def publish(fileMapping: Seq[(os.Path, String)], artifact: Artifact, release: Boolean): Unit = {
+    publishAll(release, fileMapping -> artifact)
+  }
+
+  def publishAll(release: Boolean, artifacts: (Seq[(os.Path, String)], Artifact)*): Unit = {
+    val mappings = getArtifactMappings(signed, gpgArgs, workspace, env, artifacts)
+
+    val (_, releases) = mappings.partition(_._1.isSnapshot)
+
+    val releaseGroups = releases.groupBy(_._1.group)
+    val wd = os.pwd / "out" / "publish-central"
+    os.makeDir.all(wd)
+
+    for ((_, groupReleases) <- releaseGroups) {
+      groupReleases.foreach { case (artifact, data) =>
+        val jarFile =
+          (wd / s"${artifact.group}-${artifact.id}-${artifact.version}.jar").toIO
```

Comment:
This is just for the temporary artifact that gets uploaded to Sonatype, right?
Afaik, one could also use a `.zip` file extension, which would make it clear that this is not an ordinary JAR, but an export of a Maven package hierarchy.